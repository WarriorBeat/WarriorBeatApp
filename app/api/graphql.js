/**
 * graphql.js
 * Graphql Operations
 * api
 */

import gql from "graphql-tag"
import { graphql } from "react-apollo"

// Fetch Polls
export const FetchPolls = gql`
  query FetchPolls {
    listPolls {
      items {
        pollId
        question
        date
        total_votes
        status
        answers {
          answerId
          answer
          votes
        }
      }
    }
  }
`

export const operations = {
  FetchPolls: graphql(FetchPolls, {
    options: {
      fetchPolicy: "network-only",
    },
    props: ({ data }) => ({
      loading: data.loading,
      polls: data.listPolls.items,
    }),
  }),
}
