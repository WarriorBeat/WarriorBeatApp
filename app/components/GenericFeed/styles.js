/**
 * GenericFeed/styles.js
 * Generic Feed Layout
 * Styles
 */

import { StyleSheet } from "react-native"
import { colors, soft_colors } from "config/styles"
import {
  heightPercentageToDP as hp,
  widthPercentageToDP as wp
} from "react-native-responsive-screen"

export const styles = StyleSheet.create({
  scroll_view: {
    flex: 1,
    backgroundColor: soft_colors.white
  },
  list_container: {
    flex: 1,
    flexGrow: 0,
    marginVertical: hp(".5%"),
    marginHorizontal: wp("5%"),
    backgroundColor: colors.transparent
  }
})
