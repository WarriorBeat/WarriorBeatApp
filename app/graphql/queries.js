/**
 * queries.js
 * Graphql Queries
 * graphql
 */

import gql from "graphql-tag"
import { graphql } from "react-apollo"

export const FetchPolls = gql`
  query FetchPolls {
    pollList {
      items {
        id
        question
        createdOn(format: "MMM dd yyyy")
        isOpen
        lastUpdated
        totalVotes
        options {
          id
          text
          votes
        }
      }
    }
  }
`

export const GetPoll = gql`
  query getPoll($id: ID!) {
    pollGet(id: $id) {
      id
      question
      createdOn(format: "MMM dd yyyy")
      isOpen
      lastUpdated
      totalVotes
      options {
        id
        text
        votes
      }
    }
  }
`

const queries = {
  fetchPolls: graphql(FetchPolls, {
    options: {
      fetchPolicy: "cache-and-network",
    },
    props: ({ data }) => ({
      loading: data.loading,
      polls: data.pollList ? data.pollList.items : [],
    }),
  }),
  getPoll: graphql(GetPoll, {
    options: ({ pollId }) => ({
      fetchPolicy: "network-only",
      variables: {
        id: pollId,
      },
    }),
    props: ({ data }) => ({
      loading: data.loading,
      poll: data.pollGet ? data.pollGet : {},
    }),
  }),
}

export default queries
