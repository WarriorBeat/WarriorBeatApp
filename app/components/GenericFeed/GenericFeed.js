/**
 * GenericFeed.js
 * Component to host a generic feed of posts
 * Main File
 */

import React from "react"
import { View } from "react-native"
import { observer, inject } from "mobx-react/native"
import NewsBlock from "components/NewsBlock"
import { styles } from "./styles"

@inject("postStore")
@observer
class GenericFeed extends React.Component {
  render() {
    const { categoryId } = this.props
    return (
      <View style={styles.list_container}>
        {this.props.postStore.posts.map(p => {
          if (categoryId) {
            return p.categories.find(cat => cat.id === categoryId) ? (
              <NewsBlock key={p.id} postId={p.id} />
            ) : null
          } else {
            return <NewsBlock key={p.id} postId={p.id} />
          }
        })}
      </View>
    )
  }
}

export default GenericFeed
