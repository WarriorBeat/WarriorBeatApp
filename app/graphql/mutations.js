/**
 * mutations.js
 * Graphql Mutations
 * graphql
 */

import gql from "graphql-tag"
import { graphql } from "react-apollo"
import { FetchPolls } from "./queries"

// Vote Poll
const votePollOption = gql`
  mutation pollAddVote($optionId: ID!) {
    pollAddVote(optionId: $optionId) {
      optionId
    }
  }
`

const mutations = {
  votePollOption: graphql(votePollOption, {
    props: props => ({
      onPollVote: optionId => props.mutate({
        options: {
          errorPolicy: "ignore",
        },
        variables: {
          optionId,
        },
        optimisticResponse: {
          __typename: "Mutation",
          pollAddVote: { optionId, __typename: "PollOption" },
        },
        update: (proxy, { data: { pollAddVote } }) => {
          const data = proxy.readQuery({ query: FetchPolls })
          const votedPoll = data.pollList.items.find((p) => {
            const options = p.options.map(opt => opt.id)
            if (options.includes(pollAddVote.optionId)) {
              return p
            }
            return false
          })
          const optionIndex = votedPoll.options.findIndex(
            opt => opt.id === pollAddVote.optionId,
          )
          const pollIndex = data.pollList.items.findIndex(p => p.id === votedPoll.id)
          votedPoll.options[optionIndex].votes += 1
          data.pollList.items[pollIndex] = votedPoll
          proxy.writeQuery({ query: FetchPolls, data })
        },
      }),
    }),
  }),
}

export default mutations
