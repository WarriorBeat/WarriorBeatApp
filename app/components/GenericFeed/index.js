/**
 * GenericFeed/index.js
 * GenericFeed Component
 * Index Exports
 */

import GenericFeed from "./GenericFeed"

export default GenericFeed
