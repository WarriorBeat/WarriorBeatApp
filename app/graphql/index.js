/**
 * index.js
 * Index Exports for Graphql Operations
 * graphql
 */

import { graphql } from "react-apollo"
import { graphqlMutation } from "aws-appsync-react"
import { PropTypes as types } from "prop-types"
import * as query from "./queries"
import * as mutate from "./mutations"

const pollQueries = {
  poll: {
    pollList: graphql(query.pollList, {
      options: {
        fetchPolicy: "cache-and-network",
      },
      props: ({ data }) => ({
        loading: data.loading,
        polls: data.pollList ? data.pollList.items : [],
      }),
    }),
    getPoll: graphql(query.pollGet, {
      options: ({ pollId }) => ({
        fetchPolicy: "network-only",
        variables: {
          id: pollId,
        },
      }),
      props: ({ data }) => ({
        loading: data.loading,
        poll: data.pollGet ? data.pollGet : {},
      }),
    }),
  },
}

const categoryQueries = {
  category: {
    categoryList: graphql(query.categoryList, {
      options: ({ categorySortOrder }) => ({
        fetchPolicy: "cache-and-network",
        variables: {
          sortOrder: categorySortOrder,
        },
      }),
      props: ({ data }) => ({
        loading: data.loading,
        categories: data.categoryList ? data.categoryList.items : [],
      }),
    }),
  },
}

const articleQueries = {
  article: {
    articleGet: graphql(query.articleGet, {
      options: ({ post }) => ({
        fetchPolicy: "cache-and-network",
        variables: {
          id: post.id,
        },
      }),
      props: ({ data }) => ({
        loading: data.loading,
        post: data.articleGet ? data.articleGet : {},
      }),
    }),
    articleList: graphql(query.articleList, {
      options: {
        fetchPolicy: "cache-and-network",
      },
      props: ({ data }) => ({
        loading: data.loading,
        articles: data.articleList ? data.articleList.items : [],
      }),
    }),
    articleGetByCategory: graphql(query.articleGetByCategory, {
      options: ({ category }) => ({
        fetchPolicy: "cache-and-network",
        variables: {
          categoryId: category.id,
        },
      }),
      props: ({ data }) => ({
        loading: data.loading,
        articles: data.articleGetByCategory ? data.articleGetByCategory : [],
      }),
    }),
  },
}

export const queries = {
  ...pollQueries,
  ...categoryQueries,
  ...articleQueries,
}

export const mutations = {
  poll: {
    pollAddVote: graphqlMutation(mutate.pollAddVote, query.pollList, "Poll"),
  },
}

/*
  PropTypes for GraphQL Sourced Data
*/

const PollTypes = {
  poll: types.shape({
    id: types.string,
    question: types.string,
    createdOn: types.string,
    isOpen: types.bool,
    lastUpdated: types.string,
    totalVotes: types.number,
    options: types.arrayOf(
      types.shape({
        id: types.string,
        text: types.string,
        votes: types.number,
      }),
    ),
  }),
  pollOption: types.shape({
    id: types.string,
    text: types.string,
    votes: types.number,
  }),
}

const CategoryTypes = {
  category: types.shape({
    id: types.string,
    name: types.string,
  }),
}

const PostTypes = {
  author: types.shape({
    id: types.string,
    user: types.string,
    name: types.string,
    profileImage: types.shape({
      id: types.string,
      url: types.string,
    }),
    posts: types.arrayOf(types.object),
    title: types.string,
    bio: types.string,
    staffYear: types.number,
    gradeYear: types.number,
  }),
}

const MediaTypes = {
  media: types.shape({
    id: types.string,
    createdOn: types.string,
    lastUpdated: types.string,
    credits: types.string,
    caption: types.string,
    source: types.string,
    url: types.url,
    author: PostTypes.author,
  }),
}

const ArticleTypes = {
  article: types.shape({
    id: types.string,
    createdOn: types.string,
    lastUpdated: types.string,
    author: PostTypes.author,
    title: types.string,
    categories: types.arrayOf(CategoryTypes.category),
    content: types.string,
    coverImage: MediaTypes.media,
  }),
}

export const PropTypes = {
  ...PollTypes,
  ...CategoryTypes,
  ...PostTypes,
  ...ArticleTypes,
}
