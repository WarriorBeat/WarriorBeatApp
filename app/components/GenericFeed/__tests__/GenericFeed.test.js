/**
 * __tests__/GenericFeed.test.js
 * Component Test
 * tests
 */
/* eslint-disable */

import React from "react"
import { render, shallow } from "react-native-testing-library"
import * as query from "graphql/queries"
import Providers from "tests"
import GenericFeed from "../GenericFeed"

const mocks = [
  {
    request: {
      query: query.pollList,
    },
    result: {
      data: {
        pollList: {
          items: [
            {
              id: "1",
              question: "Yes or No?",
              createdOn: "Feb 12 2019",
              isOpen: true,
              lastUpdated: "2019-02-06T19:01:35.758Z",
              totalVotes: 4,
              options: [
                {
                  id: "0",
                  text: "Yes",
                  votes: 3,
                },
                {
                  id: "1",
                  text: "No",
                  votes: 1,
                },
              ],
            },
          ],
        },
      },
    },
  },
  {
    request: {
      query: query.articleGetByCategory,
      variables: {
        categoryId: "0",
      },
      result: {
        data: {
          articleGetByCategory: {
            id: "1",
            title: "A Cool Article",
            coverImage: {
              id: "1",
              url: "https://via.placeholder.com/200",
            },
            author: {
              id: "1",
              profileImage: {
                id: "2",
                url: "https://via.placeholder.com/200",
              },
            },
          },
        },
      },
    },
  },
]

test("should render correctly", () => {
  const tree = shallow(<GenericFeed />)
  expect(tree).toMatchSnapshot()
})

test("should render data", () => {
  const tree = render(
    <Providers mocks={mocks}>
      <GenericFeed willPolls category={{ id: "0" }} />
    </Providers>,
  )
  expect(tree).toMatchSnapshot()
})
