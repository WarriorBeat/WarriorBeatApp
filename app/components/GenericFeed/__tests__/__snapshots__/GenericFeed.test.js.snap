// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
Object {
  "output": <Query
    displayName="Apollo(Apollo(GenericFeed))"
    fetchPolicy="cache-and-network"
    query={
      Object {
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "name": Object {
              "kind": "Name",
              "value": "pollList",
            },
            "operation": "query",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "pollList",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "items",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "id",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "question",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [
                                Object {
                                  "kind": "Argument",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "format",
                                  },
                                  "value": Object {
                                    "block": false,
                                    "kind": "StringValue",
                                    "value": "MMM dd yyyy",
                                  },
                                },
                              ],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "createdOn",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "isOpen",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "lastUpdated",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "totalVotes",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [
                                Object {
                                  "kind": "Argument",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "userId",
                                  },
                                  "value": Object {
                                    "kind": "Variable",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "userId",
                                    },
                                  },
                                },
                              ],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "hasVoted",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "selections": Array [
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "id",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "text",
                                    },
                                    "selectionSet": undefined,
                                  },
                                ],
                              },
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "options",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "selections": Array [
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "id",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "text",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "votes",
                                    },
                                    "selectionSet": undefined,
                                  },
                                ],
                              },
                            },
                          ],
                        },
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [
              Object {
                "defaultValue": undefined,
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NonNullType",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "ID",
                    },
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "userId",
                  },
                },
              },
            ],
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 350,
          "start": 0,
        },
      }
    }
    skip={false}
    variables={
      Object {
        "userId": undefined,
      }
    }
    warnUnhandledError={true}
  >
    [Function]
  </Query>,
}
`;

exports[`should render data 1`] = `
<View
  style={
    Object {
      "backgroundColor": "transparent",
      "flex": 1,
      "flexGrow": 0,
      "marginHorizontal": 37.5,
      "marginVertical": 6.5,
    }
  }
/>
`;
